src/main/java/org/example/Person.java

```diff
@@ -0,0 +1,78 @@
+package org.example;
+
+public class Person {
+    private String firstName;
+    private String lastName;
+    private int age;
+    private Person partner;
+
+    public Person(String firstName, String lastName, int age) {
+        this.firstName = firstName;
+        this.lastName = lastName;
+        this.age = age;
+        this.partner = null;
+    }
+
+    public String getFirstName() {
+        return firstName;
+    }
+
+    public void setFirstName(String firstName) {
+        this.firstName = firstName;
+    }
+
+    public String getLastName() {
+        return lastName;
+    }
+
+    public void setLastName(String lastName) {
+        this.lastName = lastName;
+    }
+
+    public int getAge() {
+        return age;
+    }
+
+    public void setAge(int age) {
+        this.age = age;
+    }
+
+    public Person getPartner() {
+        return partner;
+    }
+
+    public void setPartner(Person partner) {
+        this.partner = partner;
+    }
+
+    public boolean isRetired() {
+        if (this instanceof Woman) {
+            return age >= 60;
+        } else {
+            return age >= 65;
+        }
+    }
+
+    public void registerPartnership(Person partner) {
+        if (this.partner == null && partner.getPartner() == null) {
+            this.partner = partner;
+            if (this instanceof Woman) {
+                this.lastName = partner.getLastName();
+            }
+            partner.registerPartnership(this); // Зареєструвати партнерство іншого об'єкта тут
+        }
+    }
+
+    public void deregisterPartnership(boolean revertLastName) {
+        if (this.partner != null) {
+            this.partner = null;
+            if (revertLastName && this instanceof Woman) {
+                this.lastName = revertLastName();
```

Comment:
You actually do nothing here: You do set lastName from method which returns this field
For working this well you can override method `revertLastName()` in `Woman` class